src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -0,0 +1,206 @@
+import com.southernstorm.noise.protocol.CipherStatePair
+import com.southernstorm.noise.protocol.DHState
+import com.southernstorm.noise.protocol.HandshakeState
+import com.southernstorm.noise.protocol.Noise
+import io.libp2p.core.P2PAbstractChannel
+import io.libp2p.core.PeerId
+import io.libp2p.core.crypto.PrivKey
+import io.libp2p.core.crypto.unmarshalPublicKey
+import io.libp2p.core.multistream.Mode
+import io.libp2p.core.multistream.ProtocolMatcher
+import io.libp2p.core.security.SecureChannel
+import io.libp2p.etc.SECURE_SESSION
+import io.libp2p.etc.events.SecureChannelFailed
+import io.libp2p.etc.events.SecureChannelInitialized
+import io.libp2p.etc.types.toByteBuf
+import io.netty.buffer.ByteBuf
+import io.netty.channel.ChannelHandlerContext
+import io.netty.channel.ChannelInboundHandlerAdapter
+import io.netty.channel.SimpleChannelInboundHandler
+import org.apache.logging.log4j.Level
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.core.config.Configurator
+import spipe.pb.Spipe
+import java.util.concurrent.CompletableFuture
+
+open class NoiseXXSecureChannel(val localKey: PrivKey, val privateKey25519: ByteArray) :
+    SecureChannel {
+    private var role: Int = HandshakeState.RESPONDER
+    private val logger = LogManager.getLogger(NoiseXXSecureChannel::class.java.name)
+    private lateinit var localDHState: DHState
+
+    private val handshakeHandlerName = "NoiseHandshake"
+
+    companion object {
+        const val protocolName = "Noise_XX_25519_ChaChaPoly_SHA256"
+        const val announce = "/noise/$protocolName/0.1.0"
+    }
+
+    override val announce = Companion.announce
+    override val matcher = ProtocolMatcher(Mode.PREFIX, name = "/noise/$protocolName/0.1.0")
+
+    init {
+        Configurator.setLevel(NoiseXXSecureChannel::class.java.name, Level.DEBUG)
+    }
+
+    fun initChannel(ch: P2PAbstractChannel): CompletableFuture<SecureChannel.Session> {
+        return initChannel(ch, "")
+    }
+
+    override fun initChannel(ch: P2PAbstractChannel, selectedProtocol: String): CompletableFuture<SecureChannel.Session> {
+        role = if (ch.isInitiator) HandshakeState.INITIATOR else HandshakeState.RESPONDER
+
+        // configure the localDHState with the private
+        // which will automatically generate the corresponding public key
+        localDHState = Noise.createDH("25519")
+        localDHState.setPrivateKey(privateKey25519, 0)
+
+        val ret = CompletableFuture<SecureChannel.Session>()
+        val resultHandler = object : ChannelInboundHandlerAdapter() {
+            override fun userEventTriggered(ctx: ChannelHandlerContext, evt: Any) {
+                when (evt) {
+                    is SecureChannelInitialized -> {
+                        ctx.channel().attr(SECURE_SESSION).set(evt.session)
+                        ret.complete(evt.session)
+                        ctx.pipeline().remove(handshakeHandlerName)
+                        ctx.pipeline().remove(this)
```

Comment:
Sorry, I missed where the messages ciphering happens ?

Reply:
At the moment, it's taking place outside of this class in a wrapping handler. I'll move it into this class.

Reply:
done